### ISSUE TITLE
Configurable output folder

### ISSUE THREAD

**Issue by just-boris** [gatsbyjs/gatsby#1878]
Currently the project is always generating content into `public` folder. It is fine for most cases, however some people have different project structure and want to change it.

How do you think about making this configurable via CLI argument or `gatsby-config.js`? For example:
```
gatsby build --output-dir my-static-website
```

`public` will remain default, of course.

**Comment by KyleAMathews** [comment]
Do you have a specific use case in mind? This has been discussed before but no one has come up with a concrete use case that justified adding a new option.

Every option we add to Gatsby makes the project more complex which has all sorts of long-term costs so unless something is really valuable, I'd rather people handle this sort of thing themselves e.g. just copy the files to the output directory you want or create a symlink. This could easily be turned into a plugin that people could install, etc.

**Comment by just-boris** [comment]
Yes, I have a use-case. I am going to use Gatsby for a documentation part as a part of complex project. All static files (Gatsby output, plus some others) should be placed into one folder `build`, that will be deployed somehow later. In other words, the Gatsby output is only one subfolder in my setup.

So far I have worked this around in `postbuild` step, but it looks hacky:

```
"build": "gatsby build",
"postbuild": "mv public build/gatsby-subsite"
```
Adding configurable output folder will reduce this complexity and will help me not to move files around one more time.

**Comment by KyleAMathews** [comment]
Why do outputted files need to be in a folder named "build"?

**Comment by just-boris** [comment]
Because I am composing the final result not only from Gatsby output, but also from some other sources. Gatsby will be responsible only for `gatsby-subsite` folder in the tree bellow:

```
└── build
    ├── assets
    ├── gatsby-subsite
    │   └── generated-website-here...
    └── index.html
```

**Comment by domparry** [comment]
I have a similar requirement, but I just use the CI tool to rename/move the folder as a post build step... I wouldn't say it's a gatsby concern this. Feels very CI to me.

**Comment by just-boris** [comment]
For the time being, I have to do the same, that I have mentioned in the [my comment above](https://github.com/gatsbyjs/gatsby/issues/1878#issuecomment-324074853). But there are some downsides:

* I need to keep in mind, that there is a temporary `public` folder, that has to be present in `.gitignore`
* It requires an extra step of moving files around.
* Recently added feature https://github.com/gatsbyjs/gatsby/issues/1811 obviously will not work, because files will be moved under a new location.

**Comment by KyleAMathews** [comment]
The general rule of thumb is only put stuff in core that can only be done in core and since this is more trivial to do in userland than core, I'm marking this #wontfix. 

I'd love to see a plugin that implements a very efficient sync step to another directory e.g. default to rsync if available and use a node rsync-like implementation.

Re: #1811 rsync makes it easy to sync deletes as well.

**Comment by monsieurnebo** [comment]
@KyleAMathews  We use a generic continuous integration configuration for all our projects, based on a `/dist` directory. It's unfortunate that a script is needed (for Gatsby based) projects in order rename the output folder instead of a useful Gatsby option...

Any way of reconsider this feature?

**Comment by willdady** [comment]
Seems like a pretty trivial addition many people would find useful IHMO. [Github pages](https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/) expects the output to be in the root of the repository or in a `docs/` directory.

**Comment by KyleAMathews** [comment]
Ok — I guess it is annoying enough / inefficient enough to have to copy files around to add this as an option.

Who wants to own the change? Public is hardcoded a number of places in the codebase. Also the new option would need documented, etc.

**Comment by merlinpatt** [comment]
What's the update on this issue? 

It looks like there are a couple pull requests ready but not merged yet. 

What else do the PRs need to be ready?

**Comment by m-allanson** [comment]
Done in #4756. Thanks @Calderis!

**Comment by m-allanson** [comment]
#4756 had a few issues and was reverted. Reopening this.

**Comment by merlinpatt** [comment]
What issues did it have?

**Comment by m-allanson** [comment]
`build` and `develop` weren't working in many cases, see the referenced issues in https://github.com/gatsbyjs/gatsby/pull/4756

**Comment by trickydisco78** [comment]
What a shame. I really need to change the output paths as we are hosting a one off page using gatsby and been told by the apache guys they need the file in folders so they know which one's to proxy

(we've got a complicated set up and i'm asking them to take the main homepage out of the cms and host a single page which is proxied to look like it's come from the root)

I've tweaked the page now that i'm getting 99-100% on google lighthouse!

**Comment by trickydisco78** [comment]
The other option. Is there anyway of prefixing the output js and css files?

the server team need a way to proxy these files to the root as it's going to be the homepage. They need a way of identifying requests for gatsby specific js files. Can i append something to the chunkname?

**Comment by grega** [comment]
@trickydisco78 we've got around this using:
```
gatsby build --prefix-paths
```

and setting the following in `gatsby-config.js`:
```
pathPrefix: '/homepage-9df4b',
```

(the `9df4b` identifier is just to ensure the path is unique)

This results in request paths such as `/homepage-9df4b/app-fd734074ed92697dcac6.js` which we can then route to the correct host (we use a Cloudflare Worker to route requests for the homepage / homepage assets to a Gatsby site on Netlify, whilst the rest of the site is served up on other infrastructure).

It's a bit of a faff (hence our interest in the above), but it does work.

**Comment by grega** [comment]
> I see that that changes the requests not that actual build

Yep exactly that! (sorry wasn't completely clear in my original post)

**Comment by monsieurnebo** [comment]
@tribou The path prefix thing **wont create** the folder, it will make Gatsby **works in this folder** (that you have to create the way you want).

**Comment by trickydisco78** [comment]
Ah cool thanks.

Did also wonder about prefixing the filenames listed in the webpack config

`filename: `[name]-[contenthash].js`,
 chunkFilename: `[name]-[contenthash].js`,`

Any downsides of doing this (apart from having to support it with a postinstall patch)

**Comment by cromosom** [comment]
@grega Won't this prefix your url paths too? like `domain.com/page/name` -> `domain.com/prefix/page/name`

How would I avoid this?

**Comment by trickydisco78** [comment]
I've set this up and it's working great apart from all the content lives in a subfolder (on apache) and the server has been set up to serve content at root from this folder /homepage

Problem is it's now redirecting to the folder

Example: http://wwwpreprod.bristol.ac.uk

redirects to http://wwwpreprod.bristol.ac.uk/homepage-preprod

Any idea why?

**Comment by trickydisco78** [comment]
IF you switch JS off the redirect doesn't occur. In dev tools i can't see this happening. Is there somethign i can change with gatsby to prevent this?

**Comment by freshcoat** [comment]
Hello,

My scenario is the needs of having one code base but to ability to build different production builds with different languages.

It would be pity if such option won't be supported as I want to pull website translations based on a env locale variable through i18next during the production build. What I don't want is a website based on a language switcher, I don't want to put any locale parameter in my urls like: '/:locale/slug-name'.

What I would like to be able to is something like this:
```
public
├── uk
│   └── generated-uk-website-here...
├── es
│   └── generated-es-website-here...
├── fr
│   └── generated-fr-website-here...
```

And then set the correct website language version per domain by .htaccess.

So, if this will be impossible. What solution would you provide to achieve my needs? A code base for +10 websites won't be a solution of course.

**Comment by paulmelnikow** [comment]
I'm working on https://shields.io/, which has a relatively complicated setup involving two servers in a monorepo. There is an existing `public/` folder which serves a different purpose today and I'd like to build into `build/` to avoid conflict with the existing `public/` and to make a gentler migration path.

There are many use cases where this provides significant utility. In our case it's not a matter that can be solved with a simple rename.

**Comment by skerschb** [comment]
We need to support parallel builds in a kubernetes environment where we do not have control over how many gatsby builds are running in the same pod -- and against the same volume. The builds will clobber each other. I agree that most of these other use cases can be handled out of core, but in our case, we need to make sure the builds are outputted to separate directories.

**Comment by dlimx** [comment]
Does anyone in the community want to revisit #4756 together? 

We're also stuck at the JS based redirects to a subroute - we're considering either looking to fork a configurable output or dropping JS redirects.

We'd really like to see this issue reopened in some way, even if it's just to keep discussion open.

**Comment by tjmcewan** [comment]
I have a use case.  In the same vein as https://github.com/gatsbyjs/gatsby/issues/4512 and given the differences between prod builds & dev builds, I need to re-run prod builds for checking global style conflicts or omissions.  More pressing than this, but also much more niche, I'm using gatsby to automate the creation of many print pdfs that share data with our website.  The dev server output is unsuitable for PrinceXML, so to check the pdf generation I need to stop the dev server, run the build, generate the pdfs, then start the dev server again.  I can't just copy the output to another directory because running build in another session breaks the dev server.

**Comment by tyoc213** [comment]
I have a use case, currently Im testing the new https://github.com/web2py/web3py and apparently you can put some htmls there directly... but I would love to setup a submodules type of approach being the father the backend (web3py) or a gatsby site...

The setup of the example app is some like

```
$ tree
.
├── __init__.py
├── __pycache__
│   └── __init__.cpython-37.pyc
├── databases
│   ├── 452ae4793fa4a441d09c1c4fb33c04f3_thing.table
│   ├── README
│   ├── sql.log
│   └── test
├── static
│   ├── css
│   │   └── bulma.css
│   ├── hello.txt
│   └── js
│       ├── axios.min.js
│       ├── sugar.min.js
│       └── vue.min.js
└── templates
    ├── form.html
    ├── generic.html
    └── layout.html
```

And the controller is some like

```
# exposed as /examples/showme
@action('showme')
@action.uses('generic.html')
def showme():
   pass
```

So it renders with the generic.html.

Well, so the idea would be to have the web3py static pages generated from the gatsby project... at less that is my idea now.

**Comment by ewagstaff** [comment]
Maybe related? It's not a deal breaker, but it's pretty annoying that `gatsby develop` and `gatsby build` both create files in the `public` folder. We're using path prefixes the make sure all our files work in the directory we deploy to, but that means `gatsby build` overwrites the files in `public` with the prefix and breaks CSS/JS running in another terminal tab via `gatsby develop`.  The terminal window indicates no compile errors, but localhost is completely shot. 

The result of this is that every time we deploy, it requires that we stop and restart our development tab. Again, not a deal breaker but it does get tiresome when you're deploying dozens of times with small changes.

So yeah, if `gatsby develop` could run from files in a new `develop` folder and `gatsby build` could create `public`, that would be ideal.

**Comment by coltonbh** [comment]
I'd like to add a new use case for this feature!

We current use gatsby to build medical reports for clinic members. Our input to the build process is a directory containing files with arrays of JSON objects and images/movies. We use gatsby-source-filesystem to ingest these files and then build a unique static site for each member who visits our facility that summarizes their medical data from a visit. Gatsby is great for this!

I'm starting to build a more serious pipeline for automating gatsby site generation. Using AWS Lambda to build gatsby sites would be ideal (inject this asset package from S3, build report on Lambda, publish to a new bucket that serves the static site); however, since AWS Lambda only allows writes to the /tmp/ directory, I can't build our reports on this service. Lambda seems like an ideal infrastructure piece for building and publishing sites in this situation where I'd need to build thousands of sites each from a unique data input. Additionally, the solution of writing files to /public and then copying them over using some other tool isn't as option since AWS Lambda's filesystem is read-only except for the /tmp/ directory.

In short--using AWS Lambda for building gatsby sites is not an option. Being able to build to a specific directory would make this a possibility and it would be a great way for us to build thousands of gatsby sites programatically.

**Comment by trickydisco78** [comment]
@coltonbh I believe this is now in gatsby. You can specify an asset prefix

https://www.gatsbyjs.org/docs/path-prefix/

`The assetPrefix feature can be thought of as semi-related to this feature. That feature allows your assets (non-HTML files, e.g. images, JavaScript, etc.) to be hosted on a separate domain, for example a CDN.

This feature works seamlessly with assetPrefix. Build out your application with the --prefix-paths flag and you’ll be well on your way to hosting an application with its assets hosted on a CDN, and its core functionality available behind a path prefix.`

**Comment by paulmelnikow** [comment]
Ah, that's something different. It's for building a site so it can be hosted somewhere other than the root of the domain.

This issue is about outputting the build site to a different place on disk than `public`.

**Comment by coltonbh** [comment]
@paulmelnikow you are correct, the --prefix-paths would not solve my problem at all.

I need to generate one gatsby site per data assets folder (I have thousands of different asset folders) and then upload the built site (typically '/public') to an S3 bucket where we host the sites.

Of course I could move to AWS Batch or other more heavy weight job engines for doing this, but Lambda would be such a nice service and it's a shame to need a heavy weight job engine only because I can't build a gatsby site to a path other than './public'. Using Lambda S3 triggers to build and publish a new site every time we upload a new assets package would also be fantastic.

Please keep this feature request alive! :)

**Comment by MpMeetPatel** [comment]
**"postbuild": "cp -r public/* build/", --> you can have any name intead of build folder ofcourse.**

> make sure directory build exists before you run the command

I use this command (The main reason of needing this , i put public folder into .gitignore and i only push build code on github/gitlab)

**Comment by KyleAMathews** [comment]
@coltonbh can you install Gatsby inside the `/tmp` folder?

**Comment by coltonbh** [comment]
@KyleAMathews good idea! I already went down that route though. The /tmp/ directory has a limit of 512MB storage. Once I install the gatsby site with all it's associated NPM modules we're up to about 350-400MB in the /tmp/ dir, then I add our assets directory that we build from (+75MB) and then after building the site I have an additional ~75MB of files in the ./public folder and I'm either right up against or slightly over the allowed 512MB of storage space. It could work for some builds, but for others it would fail depending on the total build size, so it can't reliably work as part of our build pipeline.

Do you have any other suggestions/best practices for building thousands of gatsby sites? Our use case is a bit strange since most people are building one website from one set of assets so typical CI tools make sense; however, we've found gatsby is great for creating these dashboard views to summarize each visit a member makes to our clinic--hence the need to generate so many sites programatically.

Also, our entire member dashboard summarizing all their health data is built on gatsby--people love the reports and gatsby is a phenomenal tool for generating them. Thanks for building such a great framework!

**Comment by coltonbh** [comment]
@MpMeetPatel that would not work for my use case. AWS Lambda is a read-only filesystem except for the /tmp/ directory. So building to the ./public folder and then copying over the directory to the /tmp/ directory is not possible. AWS would reject the initial writes to ./public.

**Comment by MpMeetPatel** [comment]
@coltonbh , you're right. We can't do anything in that case. so, gasby need to consider this as a feature request that may or may not be implemented by gatsby in future.

**Comment by merlinpatt** [comment]
What if, as a short term solution, Gatsby created a wraparound script that would be activated by a flag (`--buildDirectory=something`) that just changes the directory name after it's been created?

**Comment by KyleAMathews** [comment]
Can you create symlinks on lambdas? Make a symlink from the tmp directory
to the public folder?

**Comment by coltonbh** [comment]
@KyleAMathews another good idea! But you cannot create any entry to the Lambda filesystem outside of the /tmp/ directory--including symlinks. In fact, it looks like the "ln -s" command isn't even available in the Lambda execution environment--the file system access is very limited and confined to the /tmp/ dir.

**Comment by muescha** [comment]
Ich looked over the sources for `public`. The biggest problem is that everyone construct the `public` dir hardcoded. 

If we move it to a core helper function which give everyone the output dir then it is possible to tweak the output dir. 

It is a long process to switch this also for plugins. But we should start to have only one single point of truth for this output dir.

**Comment by alexluong** [comment]
Hey @mathieudutour I know you have some experiences with this issue. Any chance you can give some thoughts on this?

**Comment by mathieudutour** [comment]
Not sure what you want me thoughts on, but I worked around the issue by patching gatsby: https://github.com/airbnb/Lona/tree/master/compiler/generate-docs/patches

It's really less than ideal and I already submitted a PR for this. The codebase is using hardcoded paths in a lot of places and even if we don't want to make them configurable, they should come from a single helper function anyway. Leaving it like this is only going to make the issue worse (because maintainers are going to hardcode it even more places and the maintenance cost is increasing).

**Comment by coltonbh** [comment]
I agree with @mathieudutour that having a single helper function that defines these paths is probably a good way to go :)

**Comment by wardpeet** [comment]
https://github.com/gatsbyjs/gatsby/issues/14341#issuecomment-496846091

> More and more people might need this so we might consider this for v3. I'll discuss with the team.

**Comment by muescha** [comment]
i would like also to vote for this idea

> @Ronva @lannonbr 
> I've not looked into the details of gatsby right now. I guess its a lot more work to do than to define interfaces and we are good to go. The examples above would require some sort of filesystem layer which could be replaced. The from my point of view the outine would similar to:
> 
> - Identify ideas and usecases solvable by these output plugins
> - Look through sources where these plugins would need to be integrated (where are file accesses to '.cache' or 'public')
> - Define common interface for plugins
> - Move file system output (.cache and public folders) into a default output plugin
> 
> Is this what you need to hear from me, or did I misunderstood you?
> 

_Originally posted by @KnisterPeter in https://github.com/gatsbyjs/gatsby/issues/14341#issuecomment-496464447_

and also add that all path generation/calculation (for read and write) should be done by this core helper and it should be with hooks and some replacable plugins, so it can be tweaked for each implementation.

then also a `pathPrefix` can be implemented with this central helper and it not need to expose all the `prefixPath` method the way down to all other methods as an parameter.

**Comment by emileswain** [comment]
so let me get this straight, i can't change the output directory for my build ? Like nearly every build process in the world allows you to do this. And to boot, you also build the dev build to the same directory as the prod build ? So when running locally, i essentially have to now concern myself with possible conflicts between prod and dev files conflicting? 

bewildered.

**Comment by mglaman** [comment]
One reason I would like this:

* Single GatsbyJS app
* Source is a content repository supporting different markets, languages
* Data in the environment control the GatsbyJS source
* Controlling output-dir allows building many sites off of one GatsbyJS app

**Comment by muescha** [comment]
FYI: 

in https://github.com/gatsbyjs/gatsby/issues/12817#issue-424675313 is a link to this gist https://gist.github.com/digitalkaoz/94933c246ba67032a1507083e2605a30
which shows a mock file system for the paths `.cache` and `public`. but it is a little bit unclear if it is just an idea or an working solution.

**Comment by mathieudutour** [comment]
> a mock file system for the paths .cache and public

I tried that and it's not that simple. It might have been at some point but not anymore. IIRC it's because gatsby uses workers that shell out another instance of Node and the mock is not working there.

**Comment by digitalkaoz** [comment]
> > a mock file system for the paths .cache and public
> 
> I tried that and it's not that simple. It might have been at some point but not anymore. IIRC it's because gatsby uses workers that shell out another instance of Node and the mock is not working there.

It worked when i used gatsby within AWS Lambda at some point. But the whole thing was that messy, that i switched out gatsby with react-static

**Comment by moroshko** [comment]
It's 2020, and we still don't have a clean solution for custom build folder.

Here is my use case:

I build a Design System and have its docs as a Gatsby site residing in the same repo.

```
website
  src
    pages
      index.js
      ...
  package.json
src
  components
    button
    text
    ...
  index.js
package.json
```

I use ZEIT's [now](https://zeit.co) to deploy the Gatsby site. 
The problem is that, ZEIT's GitHub integration is expecting a `public` folder at the root of the repo (otherwise, you get [an error](https://zeit.co/docs/v2/platform/frequently-asked-questions#missing-public-directory)).

Sounds like the recommended approach for now is to add a [postbuild script](https://github.com/gatsbyjs/gatsby/issues/1878#issuecomment-324074853).

Or, is there a better way? @KyleAMathews @rauchg

**Comment by yugasun** [comment]
@moroshko There is still no solution, this is [hard code](https://github.com/gatsbyjs/gatsby/blob/f4386bcd955316f65eca175e5147f20cb0b70cd0/packages/gatsby/src/utils/webpack.config.js#L149), so I just add `mv public dist` after `gatsby build`.🤷‍♀️🤷‍♀️🤷‍♀️

**Comment by akkie** [comment]
We try to implement a catalog builder that builds static product catalogs for different customers. The catalog builder is able to build catalogs in parallel. So we need different output paths for every build.

**Comment by benedictjohannes** [comment]
I would vouch the benefits of configurable output folder. Easiest use case is #4512 (differentiating dev and production build destination), very useful when opening production result in one browser tab (via `gatsby serve` or statically using localhost Apache/Nginx and trying out development build in another tab.

**Comment by pavelloz** [comment]
It would be great if there was an option to control output for at least html vs assets files.
I think it would be helpful to keep assets like css/js/fonts in a separate directory inside `public/` folder.
They often are moved to CDN (at least in my experience) and it would be much easier to have them grouped + properly configured urls with assetPrefix. This would remove a lot of post-build scripting and probably make intergations with a lot of platforms easier.

**Comment by hostingnuggets** [comment]
Any estimated time on when this feature will be implemented? I think it would be really useful.

**Comment by bradley** [comment]
I would also like this. My use case is that some clients want a Sanity studio client with Gatsby for the web app, where I run both through Netlify (e.g.; Netlify actually manages both the Sanity studio and Gatsby web app at a single URL rather than the more basic approach where they have separate domains). In order to make this a sane developer experience, the Sanity and Gatsby apps exist as separate sub folders in a `lib` directory, and both are built to a `dist` directory for production:

```
app/
  dist/          # gitignored and built for production
    studio/
    web/
  lib/            # development source code in version control
    studio/
    web/
  package.json
  ...etc
```

**Comment by stevemarksd** [comment]
I have another use case:

I have multiple websites that derive from a main website. All that changes is a google analytics. I could copy paste the whole project, but doesn't make sense. So currently I need to deploy each website sequentially. Another problem is that keeps accumulating js files from variant A to variant B... so variant X will have all js from previous variants. 

Would be cool to output to `public/variantA`, `public/variantB`.... especially when the only thing that changes is a small string. 

They all get deployed to a different domain.

**Comment by seht** [comment]
> Do you have a specific use case in mind? This has been discussed before but no one has come up with a concrete use case that justified adding a new option.
> 
> Every option we add to Gatsby makes the project more complex which has all sorts of long-term costs so unless something is really valuable, I'd rather people handle this sort of thing themselves e.g. just copy the files to the output directory you want or create a symlink. This could easily be turned into a plugin that people could install, etc.

I mean, this answer - what kind of argument is that? Spreading every asset onto the root folder is simply unthinkable let alone . I can think of several reasons why this is a bad practice... But the one that's briging me here is regex for cdn caching eg firebase... So where is this at? This can easily be done with webpack...

'static' here would be the prefix...

```JS
    output: {
        filename: 'static/js/[name].[chunkhash].js',
        chunkFilename: 'static/js/[name].[chunkhash].js',
        path: path.resolve(__dirname, 'public'),
        publicPath: '/',
    },
    optimization: {
        splitChunks: {
            chunks: 'all',
        }
    },
```

```JS
        new MiniCssExtractPlugin({
            filename: 'static/css/[name].[chunkhash].css',
            chunkFilename: 'static/css/[name].[chunkhash].css',
        }),
```

Since it appears that you're already using ```outputPath: 'static'``` for the file-loader... (asset files ref from css like fonts...) it would only make it consistent. In fact you could argue that the folder could be named "static" and not make it configurable. But this root folder mess... XD

Anyway, let me know -- I can help looking into this. 

Don't mind helping out..

**Comment by digitalkaoz** [comment]
All discussions about simply moving things around in bash doesnt solve the Problem running gatsby in a readonly Filesystem (Lambda). so gatsby is simply not usable within cloudfunctions: so a big nope for gatsby. I recommend react-static (where it was clean to implement and contribute)

**Comment by mzedeler** [comment]
My project is using automated build on github followed by publishing to github pages. The folder name on github is `docs`, so we have to move things, which is the current workaround. I just find it peculiar and an obvious code smell that such a basic thing can't be implemented easily. At the same time, I do acknowledge that such a change may require a lot of work due to the way the project was conceived to begin with. I just hope that someone can fix it at some point in the future.

**Comment by seht** [comment]
I think react-static is exactly what I was looking for.

Surely there must be a reason. But unfortunately somthing like this won't work with gatsby..

```JS
const path = require("path")

exports.onCreateWebpackConfig = ({
    stage,
    rules,
    loaders,
    plugins,
    actions,
}) => {
    actions.setWebpackConfig({
        output: {
            filename: 'static/js/[name].[chunkhash].js',
            chunkFilename: 'static/js/[name].[chunkhash].js',
            path: path.resolve(__dirname, 'public'),
            publicPath: '/',
        },
    })
}
```

--> https://www.gatsbyjs.com/docs/add-custom-webpack-config/

-> setWebpackConfig would be the place to look how these options are possibly being overwritten.. would be worth looking into. (perhaps a deep diff) Imma try react-static - thanks! But this has nothing to do with SSR or functions.. hmm

**Comment by seht** [comment]
According to RTFM something like this won't work with gatsby...

```JS

exports.onCreateWebpackConfig = ({
    stage,
    rules,
    loaders,
    plugins,
    actions,
}) => {
    actions.setWebpackConfig({
        output: {
            filename: `${static}/js/[name].[chunkhash].js`, // this pattern should be abstracted so that it can be applied to plugins (bellow)
            chunkFilename: '${static}/js/[name].[chunkhash].js',
            // path: path.resolve(__dirname, 'public'),
            publicPath: '/', // where pathPrefix would go...
        },
    })
}
```
```JS
        new MiniCssExtractPlugin({
            filename: '${static}/css/[name].[chunkhash].css',
            chunkFilename: '${static}/css/[name].[chunkhash].css',
        }),
```
example loaders...
```JS
{
                test: /\.s?css$/,
                use: [
                    {
                        loader: (isProd) ? MiniCssExtractPlugin.loader : 'style-loader',
                    },
                    {
                        loader: 'css-loader',
                        options: {
                            modules: false,
                            url: true,
                        }
                    },
                    {
                        loader: 'sass-loader',
                    },
                ],
            },
            {
                test: /\.(svg|eot|ttf|woff(2?))$/,
                loader: 'file-loader',
                options: {
                    name: '[name].[contenthash:20].[ext]',
                    outputPath: 'static/fonts',
                },
            },
            {
                test: /\.(png|jpe?g|gif)$/,
                loader: 'file-loader',
                options: {
                    name: '[name].[contenthash:20].[ext]',
                    outputPath: 'static/images',
                },
            },
```

--> https://www.gatsbyjs.com/docs/add-custom-webpack-config/

-> setWebpackConfig left it to https://github.com/survivejs/webpack-merge rightly so perhaps, but not entirely... -> I'm not sure yet I could get away w building a plugin for this since the only one available is not cutting cut the mustard.. So I might just get my hands dirty and fix this mess. XD I mean this is the solution... if anyone wants to grab it rn pls!

**Comment by seht** [comment]
According to RTFM something like this won't work with gatsby...

--> https://www.gatsbyjs.com/docs/add-custom-webpack-config/

```JS
// const assetsPrefix = 'static'
exports.onCreateWebpackConfig = ({
    stage,
    rules,
    loaders,
    plugins,
    actions,
}) => {
    actions.setWebpackConfig({
        output: {
 // this pattern should be abstracted so that it can be applied to plugins (bellow)
            filename: `${assetsPrefix}/js/[name].[chunkhash].js`,
            chunkFilename: '${assetsPrefix}/js/[name].[chunkhash].js',
            // path: path.resolve(__dirname, 'public'),
            publicPath: '/', // where pathPrefix would go...
// i.e. the solution would be adding a new directive in this context 'assetsPrefix'.
        },
    })
}
```
Example of a webpack config continued...
```JS
{
plugins: [
        new MiniCssExtractPlugin({
            filename: '${assetsPrefix}/css/[name].[chunkhash].css',
            chunkFilename: '${assetsPrefix}/css/[name].[chunkhash].css',
        }),
```
example loaders...
```JS
{
                test: /\.s?css$/,
                use: [
                    {
                        loader: (isProd) ? MiniCssExtractPlugin.loader : 'style-loader',
                    },
                    {
                        loader: 'css-loader',
                        options: {
                            modules: false,
                            url: true,
                        }
                    },
                    {
                        loader: 'sass-loader',
                    },
                ],
            },
            {
                test: /\.(svg|eot|ttf|woff(2?))$/,
                loader: 'file-loader',
                options: {
                    name: '[name].[contenthash:20].[ext]',
                    outputPath: '${assetsPrefix}/fonts',
                },
            },
            {
                test: /\.(png|jpe?g|gif)$/,
                loader: 'file-loader',
                options: {
                    name: '[name].[contenthash:20].[ext]',
                    outputPath: '${assetsPrefix}/images',
                },
            },
```

-> setWebpackConfig left it to https://github.com/survivejs/webpack-merge rightly so perhaps; but this would be a solution.

**Comment by pavelloz** [comment]
I guess we all agree :)

**Comment by danyim** [comment]
> It would be great if there was an option to control output for at least html vs assets files.
I think it would be helpful to keep assets like css/js/fonts etc. (generally, referenced from html, webpack dynamic imports, etc.) in a separate directory inside public/ folder.

> They often are moved to CDN (at least in my experience) and it would be much easier to have them grouped + properly configured urls with assetPrefix. This would remove a lot of post-build scripting and probably make intergations with a lot of platforms easier.

I want to second this feature. Currently the output directory has a mix of cache-busting JS/CSS files and HTML, which makes it really hard to separate manually with scripts. For my particular use case, I deploy Gatsby via FTP, and when updating the site, it would be convenient to have a configurable output directory structure so that I know which directories to replace.
